readme.md: example code does not compile

```sh
./iago_test.go:16:34: not enough arguments in call to iagotest.CreateSSHGroup
	have (*testing.T, number)
	want (*testing.T, int, bool)
./iago_test.go:18:8: not enough arguments in call to g.Run
	have (unknown type)
	want (string, func(ctx context.Context, host iago.Host) error)
./iago_test.go:18:13: undefined: iago.Task
./iago_test.go:21:15: undefined: iago.P
./iago_test.go:22:15: undefined: iago.P
./iago_test.go:23:4: unknown field Mode in struct literal of type iago.Download
```

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -18,6 +18,7 @@
         "kevinburke",
         "knownhosts",
         "mrand",
+        "Println",
         "relab",
         "sftpfs",
         "testdata",

--- a/README.md
+++ b/README.md
@@ -1,63 +1,76 @@
 # Iago
 
-Iago (**I**nfrastructure **A**s **GO**) is an experimental software deployment framework.
-Iago scripts are written in Go. This means that Iago scripts can be compiled into a simple binary with no dependencies.
+Iago (Infrastructure as Code) is an software deployment framework.
+Iago scripts are written in Go.
+This means that Iago scripts can be compiled into a simple binary with no dependencies.
 
 ## Basic API
 
-Iago executes *tasks* on *groups* of *hosts*.
-Tasks describe *actions* to be performed to each individual host, and how to handle errors.
+Iago executes *tasks* on a *group* of *hosts*.
+Tasks are functions that describe the *actions* to be performed on each individual host concurrently.
+The `iago.DialSSH` or `iago.NewSSHGroup` can be used to connect to existing SSH servers.
 
 ```go
-  // var g iago.Group; this is a set of iago.Host instances that the task will be applied to.
-  g.Run(iago.Task{
-    Name: "Example Task",
-    // Here we specify the iago.Func action, which is an action that executes the function that we give it.
-    // The function is executed concurrently for each host in the group.
-    Action: iago.Func(func(ctx context.Context, host iago.Host) error {
-      log.Println(host.Name())
-      return nil
-    }),
-    // The OnError handler decides how any errors returned by the actions should be handled.
-    // Any errors are automatically wrapped in a iago.TaskError before they are given to the handler.
-    OnError: iago.Panic,
-  })
+hosts := []string{"host1", "host2", "host3"}
+configPath := "/path/to/ssh/config"
+
+// var g iago.Group is a set of iago.Host instances that the task will be applied to.
+g, err := iago.NewSSHGroup(hosts, configPath)
+if err != nil {
+	// handle error
+}
+defer g.Close()
+
+g.Run("Example Task", func(ctx context.Context, host iago.Host) error {
+  // This function will be executed concurrently on each host in the group.
+  // The context is used to control the execution of the task.
+  log.Println(host.Name())
+  return nil
+})
 ```
 
 We currently support connecting to remote hosts via SSH,
 but support for other connection methods can be added by implementing the `iago.Host` interface.
 
-### Example
-
-The following example downloads a file from each remote host.
-This example uses the `iagotest` package, which spawns docker containers and connects to them with SSH for testing.
-The `iago.DialSSH` or `iago.NewSSHGroup` can be used to connect to existing SSH servers.
+Error handling is configured at the group level, using the `ErrorHandler` field of the `iago.Group` struct.
+By default, errors cause a panic, but you can set a custom error handler:
 
 ```go
-func TestIago(t *testing.T) {
-  dir, _ := os.Getwd()
-
-  // The iagotest package provides a helper function that automatically
-  // builds and starts docker containers with an exposed SSH port for testing.
-  g := iagotest.CreateSSHGroup(t, 4)
-
-  g.Run(iago.Task{
-    Name: "Download files",
-    Action: iago.Download{
-      Src:  iago.P("/etc/os-release"),
-      Dest: iago.P("os").RelativeTo(dir),
-      Mode: 0644,
-    },
-    OnError: iago.Panic,
-  })
+g.ErrorHandler = func(e error) {
+  log.Printf("Task failed: %v", e)
 }
 ```
 
-## TODO
+### Example
 
-* Improve the reading of `ssh_config` files
-  * Should support reading the public key files (it doesn't look like the `ssher` package supports this fully)
-* Prompt the user when verifying host keys?
-* Improve logging
-  * Logging for tasks as they are running
-  * Progress -- how many hosts have finished a task
+The following example downloads a file from each remote host.
+The file is downloaded to a temporary directory created by the test framework and named `os.<hostname>`.
+See [iago_test.go](https://github.com/relab/iago/blob/master/iago_test.go#L81) for the complete example.
+
+This example uses the `iagotest` package, which spawns docker containers and connects to them with SSH for testing.
+
+```go
+func TestIago(t *testing.T) {
+  dir := t.TempDir()
+
+  // The iagotest package provides a helper function that automatically
+  // builds and starts docker containers with an exposed SSH port for testing.
+  g := iagotest.CreateSSHGroup(t, 4, false)
+
+  g.Run("Download files", func(ctx context.Context, host iago.Host) error {
+    src, err := iago.NewPath("/etc", "os-release")
+    if err != nil {
+      return err
+    }
+    dest, err := iago.NewPath(dir, "os")
+    if err != nil {
+      return err
+    }
+    return iago.Download{
+      Src:  src,
+      Dest: dest,
+			Perm: iago.NewPerm(0o644),
+    }.Apply(ctx, host)
+  })
+}
+```

--- a/fs.go
+++ b/fs.go
@@ -175,7 +175,7 @@ func (ca copyAction) Apply(_ context.Context, host Host) (err error) {
 	}
 	dest := ca.dest.path
 	if ca.fetch {
-		// since we might be copying from multiple hosts, we will prefix the filename with the host's name.
+		// since we might be copying from multiple hosts, we add the host's name to the destination file
 		dest += "." + host.Name()
 	}
 	return copyFile(ca.src.path, dest, ca.perm, from, to)

--- a/iago_test.go
+++ b/iago_test.go
@@ -77,3 +77,48 @@ func TestIago(t *testing.T) {
 		t.Log(string(f))
 	}
 }
+
+func TestIagoDownloadExample(t *testing.T) {
+	dir := t.TempDir()
+
+	// The iagotest package provides a helper function that automatically
+	// builds and starts docker containers with an exposed SSH port for testing.
+	g := iagotest.CreateSSHGroup(t, 2, false)
+
+	g.Run("Download files", func(ctx context.Context, host iago.Host) error {
+		src, err := iago.NewPath("/etc", "os-release")
+		if err != nil {
+			return err
+		}
+		t.Logf("Downloading %s from %s", src, host.Name())
+		dest, err := iago.NewPath(dir, "os")
+		if err != nil {
+			return err
+		}
+		t.Logf("Saving to %s", dest)
+		return iago.Download{
+			Src:  src,
+			Dest: dest,
+			Perm: iago.NewPerm(0o644),
+		}.Apply(ctx, host)
+	})
+
+	filesInDir, err := os.ReadDir(dir)
+	if err != nil {
+		t.Fatal(err)
+	}
+	files := make([]string, len(filesInDir))
+	for i, file := range filesInDir {
+		files[i] = file.Name()
+	}
+	t.Logf("Copied files: %s", strings.Join(files, ", "))
+
+	for _, h := range g.Hosts {
+		t.Logf("Reading file %s for host %s", filepath.Join(dir, "os."+h.Name()), h.Name())
+		f, err := os.ReadFile(filepath.Join(dir, "os."+h.Name()))
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Log(string(f))
+	}
+}

